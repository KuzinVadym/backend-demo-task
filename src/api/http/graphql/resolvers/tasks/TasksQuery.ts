import {
  Arg,
  Ctx,
  FieldResolver,
  Int,
  Query,
  Resolver,
  Root
} from 'type-graphql';
import { withHttpService } from '../../../shared/utils/withHttpService';
import { TaskType } from './TaskType';

@Resolver((of) => TaskType)
export class TasksQuery {
  @Query(() => [TaskType])
  public async tasks(@Ctx() ctx): Promise<TaskType[]> {
    const userHttpService = withHttpService('tasks', ctx.getState);
    return userHttpService.getTasks();
  }

  @Query(() => TaskType)
  public async task(
    @Arg('id', (type) => Int) id: number,
    @Ctx() ctx
  ): Promise<TaskType> {
    const userHttpService = withHttpService('tasks', ctx.getState);
    return userHttpService.getTask(id);
  }

  @FieldResolver()
  async createdByUser(@Root() task: TaskType, @Ctx() ctx) {
    const userHttpService = withHttpService('users', ctx.getState);
    return userHttpService.getUser(task.createdByUserId);
  }

  @FieldResolver()
  async updatedByUser(@Root() task: TaskType, @Ctx() ctx) {
    const userHttpService = withHttpService('users', ctx.getState);
    return userHttpService.getUser(task.updatedByUserId);
  }

  @FieldResolver()
  async assignedUser(@Root() task: TaskType, @Ctx() ctx) {
    const userHttpService = withHttpService('users', ctx.getState);
    return userHttpService.getUser(task.assignedUserId);
  }

  @FieldResolver()
  async commits(@Root() task: TaskType, @Ctx() ctx) {
    const userHttpService = withHttpService('commits', ctx.getState);
    return userHttpService.getCommits({ taskId: task.id });
  }
}
