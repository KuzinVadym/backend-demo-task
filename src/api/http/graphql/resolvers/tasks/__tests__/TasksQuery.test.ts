import 'reflect-metadata';
import {
  ApolloServerTestClient,
  createTestClient
} from 'apollo-server-testing';
import { ApolloServer } from 'apollo-server-express';
import { buildSchema } from 'type-graphql';
import { TasksQuery } from '../TasksQuery';

const contextMock = {
  getState: jest.fn()
};

const tasksServiceMock = {
  getTasks: jest.fn(),
  getTask: jest.fn()
};

const usersServiceMock = {
  getUser: jest.fn()
};

describe('TaskQuery', () => {
  let server: ApolloServer;

  beforeAll(async () => {
    const schema = await buildSchema({ resolvers: [TasksQuery] });
    server = new ApolloServer({
      schema,
      context: contextMock
    });
  });

  beforeEach(async () => {
    tasksServiceMock.getTasks.mockResolvedValue([
      {
        id: 1,
        title: 'title',
        description: 'description',
        status: 'ToDo',
        currentUserId: 1,
        updatedUserId: 1,
        createdAt: new Date()
      }
    ]);
    tasksServiceMock.getTask.mockResolvedValue({
      id: 1,
      title: 'title',
      description: 'description',
      status: 'ToDo',
      currentUserId: 1,
      updatedUserId: 1,
      assignedUserId: 1,
      createdAt: new Date()
    });
    usersServiceMock.getUser.mockResolvedValue({
      id: 1
    });

    contextMock.getState.mockReturnValue({
      httpServices: {
        tasks: tasksServiceMock,
        users: usersServiceMock
      }
    });
  });

  it('returns tasks data', async () => {
    const testQuery = `{
      tasks {
        title
        description
        status
      }
    }
  `;

    const { query }: ApolloServerTestClient = createTestClient(server);

    const res = await query({ query: testQuery, variables: {} });
    expect(res).toMatchSnapshot();
  });

  it('returns task data', async () => {
    const testQuery = `
    query TaskQuery($id: Int!){
      task(id: $id) {
        title
        description
        status
      }
    }
  `;
    const { query }: ApolloServerTestClient = createTestClient(server);

    const res = await query({ query: testQuery, variables: { id: 1 } });
    expect(res).toMatchSnapshot();
  });

  it('cals service with right params', async () => {
    const testQuery = `
    query TaskQuery($id: Int!){
      task(id: $id) {
        title
        description
        status
        createdByUser{
          id
        }
        updatedByUser{
          id
        }
        assignedUser{
          id
        }
      }
    }
  `;
    const { query }: ApolloServerTestClient = createTestClient(server);

    await query({ query: testQuery, variables: { id: 1 } });

    expect(tasksServiceMock.getTask).toBeCalledWith(1);
    expect(usersServiceMock.getUser).toBeCalledWith(1);
    expect(usersServiceMock.getUser).toBeCalledWith(1);
    expect(usersServiceMock.getUser).toBeCalledWith(1);
  });
});
