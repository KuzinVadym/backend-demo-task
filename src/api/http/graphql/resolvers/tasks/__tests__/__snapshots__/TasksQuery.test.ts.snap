// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskQuery returns task data 1`] = `
Object {
  "data": Object {
    "task": Object {
      "description": "description",
      "status": "ToDo",
      "title": "title",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`TaskQuery returns tasks data 1`] = `
Object {
  "data": Object {
    "tasks": Array [
      Object {
        "description": "description",
        "status": "ToDo",
        "title": "title",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
